add command `r.scriptPath` which returns the path to Rscript

sharon-wang opened the pull request:
### Summary

- adds a new R extension command `r.scriptPath` which returns the Rscript path based on the last R bin path
- this is being introduced to provide a similar option to Python's `"command": "${command:python.interpreterPath} ${file}"` that can be used in tasks.json for tests

### Release Notes

#### New Features

- N/A

#### Bug Fixes

- N/A


### QA Notes

@testlabauto could you please fill out this section on how to use the option in tasks.json? :D

Sure!  Here is a sample tasks.json using r.scriptPath

```
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "Run R File",
            "command": "${command:r.scriptPath} ${file}",
            "type": "shell",
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "presentation": {
                "reveal": "always",
                "panel": "new",
                "focus": true
            }
        }
    ]
}
```
This will run the currently open file with the current Rscript.  You can test the fail condition by having no active R session.

sharon-wang committed:
add command `r.rScriptPath` which returns the RScript path

This requires that an R session was previously started in the current session.

testlabauto reviewed (dismissed):
LGTM!

jennybc reviewed (commented):
I feel like we should add the Rscript path to `RMetadataExtra`, alongside `binpath` and `homepath`, instead of doing path surgery on `binpath` in this command. I realize this sort of just moves the surgery into runtime instantiation, but it somehow feels better. And if we develop other uses for it, it will be there waiting for us.

DavisVaughan commented:
Medium to strong agreement with @jennybc

sharon-wang committed:
move rScriptPath to RMetadataExtra

sharon-wang commented:
Sounds good to me! I've now moved `rScriptPath` to `RMetadataExtra` in [dcf393](https://github.com/posit-dev/positron/pull/6309/commits/2dcf393f6c62440dcf0f7259c572d4db2a3eb65b)

testlabauto commented:
> Sounds good to me! I've now moved `rScriptPath` to `RMetadataExtra` in [2dcf393](https://github.com/posit-dev/positron/commit/2dcf393f6c62440dcf0f7259c572d4db2a3eb65b)

Verified it still works after refactor

DavisVaughan commented on extensions/positron-r/src/r-installation.ts:
```diff
@@ -20,6 +20,9 @@ export interface RMetadataExtra {
 	readonly binpath: string;
 
+	/** R's RScript path */
+	readonly rScriptPath: string;
```
```suggestion
	readonly scriptpath: string;
```

I think matching the existing naming scheme in this interface would be preferred

DavisVaughan commented on extensions/positron-r/src/provider.ts:
```diff
@@ -216,11 +221,12 @@ export async function makeMetadata(
-	// Save the R home path and binary path as extra data.
+	// Save the R home path, binary path and RScript path as extra data.
 	// Also, whether this R installation is the "current" R version.
 	const extraRuntimeData: RMetadataExtra = {
 		homepath: rInst.homepath,
 		binpath: rInst.binpath,
+		rScriptPath: rScriptPath,
```
I think I was expecting that `scriptpath` ends up as part of the `RInstallation` itself.

That way:
- you just pass it through here with very little fanfare
- we can pull it from the `RInstallation` if we need to use it elsewhere
-  `validateMetadata()` can also perform its check on `scriptpath`

jennybc replied:
IIRC the only way to get this extra metadata into the `LanguageRuntimeMetadata` is to put it in `RMetadatExtra`. That's something we had to create, in fact, in order to keep both the binpath and homepath available to ourselves later.

jennybc replied:
These `RInstallation` instances don't last very long and are just used to instantiate `positron.LanguageRuntimeMetadata`.

DavisVaughan reviewed (commented):
My main comment is that to me this feels like part of `RInstallation`, and should still be constructed even further up. It feels similar in spirit to the `binpath` and `homepath` fields that already exist on `RInstallation`, and it would be nice to centralize the logic as much as possible in `RInstallation` if possible

Maybe @jennybc feels differently, and I would defer to her if so

jennybc commented on extensions/positron-r/src/provider.ts:
```diff
@@ -190,6 +190,11 @@ export async function makeMetadata(
 		path.join('~', rInst.binpath.substring(homedir.length)) :
 		rInst.binpath;
 
+	// Create the RScript path.
+	// The RScript path is the same as the R binary path, but with the 'R' or 'R.exe' executable
+	// replaced with 'RScript' or 'RScript.exe, respectively.
+	const rScriptPath = rInst.binpath.replace(/R(\.exe)?$/, 'RScript$1');
```
No to @DavisVaughan's question.

But more importantly, I think it's `Rscript` (lower case 's'), not `RScript`.

jennybc replied:
☝️ affects a few strings in this PR.

jennybc reviewed (dismissed):
Modulo my comments, LGTM. I think the `Rscript` path *has* to be added to `RMetadataExtra` in order to get the persistence that we want.

sharon-wang committed:
update naming from RScript --> Rscript and rScriptPath --> scriptPath

sharon-wang committed:
change metadata property scriptPath --> scriptpath

sharon-wang commented:
Thank you @jennybc and @DavisVaughan for the feedback!!

I've kept the implementation the same (`scriptpath` is in `RMetadataExtra`) and renamed RScript --> Rscript and rScriptPath --> scriptpath / scriptPath.

@testlabauto For consistency, I've changed the command to `r.scriptPath` and updated the PR description example. Would you mind giving this one more check that it's still working as expected?

sharon-wang commented on extensions/positron-r/src/runtime-manager.ts:
```diff
@@ -71,6 +71,7 @@ export class RRuntimeManager implements positron.LanguageRuntimeManager {
 		if (!metadataExtra.binpath) {
 			throw new Error('R metadata is missing bin path');
 		}
+		// metadataExtra.scriptPath may not exist yet and will be constructed via makeMetadata.
```
oop one more

```suggestion
		// metadataExtra.scriptpath may not exist yet and will be constructed via makeMetadata.
```

sharon-wang committed:
oop one more

Signed-off-by: sharon <sharon-wang@users.noreply.github.com>

testlabauto commented:
> Thank you @jennybc and @DavisVaughan for the feedback!!
> 
> I've kept the implementation the same (`scriptpath` is in `RMetadataExtra`) and renamed RScript --> Rscript and rScriptPath --> scriptpath / scriptPath.
> 
> @testlabauto For consistency, I've changed the command to `r.scriptPath` and updated the PR description example. Would you mind giving this one more check that it's still working as expected?

Looks good, @sharon-wang !
